### PR TITLE
Rule to check if Tags are defined as Top-level property

### PR DESCRIPTION
**Pull request by tejaswiMinnu**
TagsAreTopLevelPropertiesOnly - new rule checks if tags are defined as a top-level property, not in the properties bag.

https://msazure.visualstudio.com/One/_workitems/edit/26429619

### CODE REVIEW

File: packages/rulesets/src/spectral/az-arm.ts [hunk 1]
```diff
@@ -731,6 +732,21 @@ const ruleset: any = {
       },
     },
 
+    // RPC Code: RPC-Put-V1-30
+    TagsAreTopLevelPropertiesOnly: {
+      rpcGuidelineCode: "RPC-Put-V1-30",
+      description: "Tags must be defined as a top-level property, not in the properties bag.",
+      severity: "error",
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/az-arm.ts:739]: Tejaswi, the intention for this rule is to not allow tags for proxy resources at all even as a top level property. So the documentation, rule name and error messages need to reflect that - "Tags should not be specified in the properties bag for proxy resources. Consider using a Tracked resource instead."

File: packages/rulesets/src/spectral/az-arm.ts [hunk 2]
```diff
@@ -731,6 +732,21 @@ const ruleset: any = {
       },
     },
 
+    // RPC Code: RPC-Put-V1-30
+    TagsAreTopLevelPropertiesOnly: {
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/az-arm.ts:736]: >TagsAreTopLevelPropertiesOnly [](http://example.com/codeflow?start=4&length=29)

TagsAreNotAllowedForProxyResources

File: packages/rulesets/src/spectral/functions/tags-are-top-level-properties-only.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import _ from "lodash"
+import { deepFindObjectKeyPath, getProperties } from "./utils"
+
+const TAGS = "tags"
+const PROPERTIES = "properties"
+const ERROR_MESSAGE = "Tags must be defined as a top-level property, not in the properties bag."
+
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/tags-are-top-level-properties-only.ts:7]: "Tags should not be specified in the properties bag for proxy resources. Consider using a Tracked resource instead."

**Review by rkmanda — CHANGES_REQUESTED** [review]
:clock1:

File: docs/tags-are-not-allowed-for-proxy-resources.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+ARM OpenAPI(swagger) specs
+
+## Related ARM Guideline Code
+
+- RPC-Put-V1-30
+
+## Description
+
+Validates that `tags` is not defined in the properties bag, but rather as a top-level property,
```
Comment (rkmanda) [inline docs/tags-are-not-allowed-for-proxy-resources.md:17]: pl remove this line

File: docs/tags-are-not-allowed-for-proxy-resources.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+`tags` should not be specified in the properties bag for proxy resources. Consider using a Tracked resource instead.
+
+## How to fix the violation
+
+Ensure that any `tags` definitions are as top-level properties, not in the properties bag.
+
```
Comment (rkmanda) [inline docs/tags-are-not-allowed-for-proxy-resources.md:23]: Either remove the 'tags' property from the properties bag or consider using a tracked resource.

File: common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-tagsAreTopLevelPropertiesOnly_2024-06-13-22-19.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft.azure/openapi-validator-rulesets",
+      "comment": "Add TagsAreNotAllowedForProxyResources rule to check if tags are defined as top-level property, not in the properties bag",
```
Comment (AkhilaIlla) [inline common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-tagsAreTopLevelPropertiesOnly_2024-06-13-22-19.json:5]: >tags are defined as top-level property, not in the properties bag [](http://example.com/codeflow?start=74&length=65)

needs to be updated too.

File: docs/rules.md [hunk 7]
```diff
@@ -1213,6 +1213,12 @@ Validates that system data is not defined in the properties bag, but rather as a
 
 Please refer to [system-data-in-properties-bag.md](./system-data-in-properties-bag.md) for details.
 
+### TagsAreNotAllowedForProxyResources
+
+Validates that `tags` is not defined in the properties bag, but rather as a top-level property.
```
Comment (AkhilaIlla) [inline docs/rules.md:1218]: >Validates that `tags` is not defined in the properties bag, but rather as a top-level prope [](http://example.com/codeflow?start=0&length=91)

same

File: packages/rulesets/src/spectral/functions/tags-are-not-allowed-for-proxy-resources.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import _ from "lodash"
+import { deepFindObjectKeyPath, getProperties } from "./utils"
+
+const TAGS = "tags"
+const PROPERTIES = "properties"
+const NestedPROPERTIES = "properties"
```
Comment (AkhilaIlla) [inline packages/rulesets/src/spectral/functions/tags-are-not-allowed-for-proxy-resources.ts:6]: >properties [](http://example.com/codeflow?start=26&length=10)

do we need this?

Comment (tejaswiMinnu) [reply packages/rulesets/src/spectral/functions/tags-are-not-allowed-for-proxy-resources.ts:6]: I can directly use properties while concatenating the path, but to make it more readable I defined here.

File: common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-tagsAreTopLevelPropertiesOnly_2024-06-13-22-19.json [hunk 9]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
```
Comment (konrad-jamrozik) [inline common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-tagsAreTopLevelPropertiesOnly_2024-06-13-22-19.json:2]: Note that you no longer need to include these changelog json files. See here:

https://github.com/Azure/azure-openapi-validator/issues/683

File: docs/tags-are-not-allowed-for-proxy-resources.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+### Valid/Good Example
+
+```json
+"definitions": {
+  "Resource": {
+    "type": "object",
+    "properties": {
+    "type": "object",
+      // tope level properties
```
Comment (konrad-jamrozik) [inline docs/tags-are-not-allowed-for-proxy-resources.md:31]:
```suggestion
      // top level properties
```

File: common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-tagsAreTopLevelPropertiesOnly_2024-06-13-22-19.json (hunk 9 above)
Comment (tejaswiMinnu) [reply common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-tagsAreTopLevelPropertiesOnly_2024-06-13-22-19.json:2]: Thanks konrad removed the file

File: common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-tagsAreTopLevelPropertiesOnly_2024-06-13-22-19.json (hunk 6 above)
Comment (tejaswiMinnu) [reply common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-tagsAreTopLevelPropertiesOnly_2024-06-13-22-19.json]: Removing this file as per Konrad's suggestion

File: docs/tags-are-not-allowed-for-proxy-resources.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+- RPC-Put-V1-30
+
+## Description
+
+`tags` should not be specified in the properties bag for proxy resources. Consider using a Tracked resource instead.
+
+## How to fix the violation
+
+Either remove the 'tags' property from the properties bag or consider using a tracked resource.
```
Comment (ravimeda) [inline docs/tags-are-not-allowed-for-proxy-resources.md:21]: nit - Use consistent case. Upper case T - 'Tracked'

File: packages/rulesets/src/spectral/functions/tags-are-not-allowed-for-proxy-resources.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+      message: ERROR_MESSAGE,
+      path: _.concat(ctx.path, PROPERTIES, TAGS),
+    })
+  }
+
+  const deepProperties = deepFindObjectKeyPath(definition.properties.properties, TAGS)
+  if (deepProperties.length > 0) {
+    errors.push({
+      message: ERROR_MESSAGE,
+      path: _.concat(ctx.path, PROPERTIES, NestedPROPERTIES, deepProperties[0]),
```
Comment (bdefoy) [inline packages/rulesets/src/spectral/functions/tags-are-not-allowed-for-proxy-resources.ts:24]: Nit: rename variable for clarity
```suggestion
  const deepPropertiesTags = deepFindObjectKeyPath(definition.properties.properties, TAGS)
  if (deepPropertiesTags.length > 0) {
    errors.push({
      message: ERROR_MESSAGE,
      path: _.concat(ctx.path, PROPERTIES, NestedPROPERTIES, deepPropertiesTags[0]),
```

File: docs/rules.md [hunk 13]
```diff
@@ -1213,6 +1213,12 @@ Validates that system data is not defined in the properties bag, but rather as a
 
 Please refer to [system-data-in-properties-bag.md](./system-data-in-properties-bag.md) for details.
 
+### TagsAreNotAllowedForProxyResources
```
Comment (bdefoy) [inline docs/rules.md:1216]: Nit: consider renaming this TagsNotAllowedInPropertiesBag since it's not only for ProxyOnly resource types

File: packages/rulesets/src/spectral/az-arm.ts [hunk 14]
```diff
@@ -731,6 +732,21 @@ const ruleset: any = {
       },
     },
 
+    // RPC Code: RPC-Put-V1-30
+    TagsAreNotAllowedForProxyResources: {
+      rpcGuidelineCode: "RPC-Put-V1-30",
+      description: "Tags should not be specified in the properties bag for proxy resources. Consider using a Tracked resource instead.",
+      severity: "error",
```
Comment (AkhilaIlla) [inline packages/rulesets/src/spectral/az-arm.ts:739]: should it be marked as a warning initially?
as its a new rule?

File: packages/rulesets/src/spectral/functions/tags-are-not-allowed-for-proxy-resources.ts (hunk 8 above)
Comment (AkhilaIlla) [reply packages/rulesets/src/spectral/functions/tags-are-not-allowed-for-proxy-resources.ts:6]: I think this is generally not considered good coding practice as it creates redundancy

File: docs/rules.md (hunk 13 above)
Comment (tejaswiMinnu) [reply docs/rules.md:1217]: Initially named this "TagsAreTopLevelPropertiesOnly" due to suggestion changed to proxy
Will change to this "TagsNotAllowedInPropertiesBag"

File: docs/rules.md [hunk 15]
```diff
@@ -1213,6 +1213,12 @@ Validates that system data is not defined in the properties bag, but rather as a
...
+
+`tags` should not be specified in the properties bag for proxy resources. Consider using a Tracked resource instead.
```
[the remark this answers is not in the document]
Comment (tejaswiMinnu) [reply docs/rules.md:1219]: Will leave as is cause of lint-line limit

File: packages/rulesets/src/spectral/functions/tags-not-allowed-in-properties-bag.ts [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+      message: ERROR_MESSAGE,
+      path: _.concat(ctx.path, PROPERTIES, TAGS),
+    })
+  }
+
+  const deepPropertiesTags = deepFindObjectKeyPath(definition.properties.properties, TAGS)
+  if (deepPropertiesTags.length > 0) {
+    errors.push({
+      message: ERROR_MESSAGE,
+      path: _.concat(ctx.path, PROPERTIES, NestedPROPERTIES, deepPropertiesTags[0]),
```
Comment (bdefoy) [inline packages/rulesets/src/spectral/functions/tags-not-allowed-in-properties-bag.ts:24]: Does this work for the path? The doc comment for `deepFindObjectKeyPath` says it should return the full path
```suggestion
      path: deepPropertiesTags,
```

Comment (tejaswiMinnu) [reply packages/rulesets/src/spectral/functions/tags-not-allowed-in-properties-bag.ts:24]: I am sending only specified properties to the method which will not return the path we are expecting deepFindObjectKeyPath(definition.properties.properties, TAGS)
I also tested the suggestion you added this is resulting empty/wrong paths

**Review by bdefoy — APPROVED** [review]

File: packages/rulesets/src/spectral/functions/tags-not-allowed-in-properties-bag.ts [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+const TAGS = "tags"
+const PROPERTIES = "properties"
+const NestedPROPERTIES = "properties"
+const ERROR_MESSAGE = "Tags should not be specified in the properties bag for proxy resources. Consider using a Tracked resource instead."
+
+export const tagsNotAllowedInPropertiesBag = (definition: any, _opts: any, ctx: any) => {
+  const properties = getProperties(definition)
+  const errors = []
+
+  if ("tags" in properties && !("location" in properties)) {
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/tags-not-allowed-in-properties-bag.ts:13]: >!("location" in properties) [](http://example.com/codeflow?start=30&length=27)

You can either generalize the logic to check for tags in properties for tracked or proxy types

OR

Keep the logic to check only for proxy types and change the name of the rule.

Right now, the rule name is tags are not allowed in properties bag but the logic checks it only for proxy resources.

Comment (tejaswiMinnu) [reply packages/rulesets/src/spectral/functions/tags-not-allowed-in-properties-bag.ts]: Changed the rule name to reflect only proxy resources.

File: packages/rulesets/src/spectral/az-arm.ts [hunk 18]
```diff
@@ -736,6 +737,21 @@ const ruleset: any = {
       },
     },
 
+    // RPC Code: RPC-Put-V1-30
+    TagsAreNotAllowedForProxyResources: {
+      rpcGuidelineCode: "RPC-Put-V1-30",
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/az-arm.ts:742]: >RPC-Put-V1-30", [](http://example.com/codeflow?start=25&length=15)

rename to RPC-Put-V1-31

File: docs/tags-are-not-allowed-for-proxy-resources.md [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+
+ARM Error
+
+## Applies to
+
+ARM OpenAPI(swagger) specs
+
+## Related ARM Guideline Code
+
+- RPC-Put-V1-30
```
Comment (rkmanda) [inline docs/tags-are-not-allowed-for-proxy-resources.md:13]: 31

**Review by rkmanda — APPROVED** [review]
:shipit:
